src/client/sound/s_mix.c

```diff
@@ -69,120 +96,112 @@ _Bool S_SpatializeChannel(s_channel_t *ch) {
 
 	ch->angle = (int16_t) (360.0 - angle) % 360;
 
-	return frac < 1.0;
+	return frac <= 1.0;
 }
 
 /**
- * @brief
+ * @brief Updates all active channels for the current frame.
  */
-void S_PlaySample(const vec3_t org, uint16_t ent_num, s_sample_t *sample, int32_t atten) {
-	s_channel_t *ch;
-	int32_t i;
+void S_MixChannels(void) {
 
-	if (!s_env.initialized)
-		return;
+	s_env.num_active_channels = 0;
 
-	if (sample && sample->media.name[0] == '*') // resolve the model-specific sample
-		sample = S_LoadModelSample(&cl.entities[ent_num].current, sample->media.name);
+	s_channel_t *ch = s_env.channels;
+	for (int32_t i = 0; i < MAX_CHANNELS; i++, ch++) {
 
-	if (!sample || !sample->chunk)
-		return;
+		if (ch->sample) {
 
-	if ((i = S_AllocChannel()) == -1)
-		return;
+			if (ch->start_time) {
+				if (ch->play.flags & S_PLAY_AUTO) {
+					const cl_entity_t *ent = &cl.entities[ch->play.entity];
+
+					if (ent->frame_num != cl.frame.frame_num
+							|| ent->current.sound != ent->prev.sound) {
 
-	ch = &s_env.channels[i];
+						Mix_FadeOutChannel(i, 250);
+					}
+				}
+			}
 
-	if (org) { // positioned sound
-		VectorCopy(org, ch->org);
-		ch->ent_num = -1;
-	} else
-		// entity sound
-		ch->ent_num = ent_num;
+			if (S_SpatializeChannel(ch)) {
+				Mix_SetPosition(i, ch->angle, ch->dist);
 
-	ch->atten = atten;
-	ch->sample = sample;
+				if (!ch->start_time) {
+					ch->start_time = cl.systime;
 
-	if (S_SpatializeChannel(ch)) {
-		Mix_SetPosition(i, ch->angle, ch->dist);
-		Mix_ChannelFinished(S_FreeChannel);
-		Mix_PlayChannel(i, ch->sample->chunk, 0);
-	} else {
-		S_FreeChannel(i);
+					if (ch->play.flags & S_PLAY_LOOP) {
+						Mix_PlayChannel(i, ch->sample->chunk, -1);
+					} else {
+						Mix_PlayChannel(i, ch->sample->chunk, 0);
+					}
+				}
+
+				s_env.num_active_channels++;
+			} else {
+				
+				if (ch->start_time) {
+					Mix_HaltChannel(i);
+				} else {
+					S_FreeChannel(i);
+				}
+			}
+		}
```

Comment:
Here's the real stinker. Mix_HaltChannel doesn't call the Mix_ChannelFinished callback if the channel was never started. So failing to free channels that were never started causes us to "leak" culled channels, and eventually this blows our available channels.